File: src/D3D11/D3D11CommandList.cpp
#include "D3D11CommandList.h"

#include "D3D11Buffer.h"
#include "D3D11Pipeline.h"
#include "D3D11Texture.h"
#include "D3D11Utils.h"
#include "../Common.h"

namespace grabs::D3D11
{
    D3D11CommandList::D3D11CommandList(ID3D11Device* device)
    {
        D3D11_CHECK_RESULT(device->CreateDeferredContext(0, &Context));
    }

    D3D11CommandList::~D3D11CommandList()
    {
        if (CommandList)
            CommandList->Release();

        Context->Release();
    }

    void D3D11CommandList::Begin()
    {
        if (CommandList)
            CommandList->Release();

        CommandList = nullptr;
    }

    void D3D11CommandList::End()
    {
        D3D11_CHECK_RESULT(Context->FinishCommandList(FALSE, &CommandList));
    }

    void D3D11CommandList::BeginRenderPass(const RenderPassDescription& info)
    {
        GS_NULL_CHECK(info.Texture);
        auto d3dTexture = dynamic_cast<D3D11Texture*>(info.Texture);
        GS_NULL_CHECK(d3dTexture->RenderTarget);

        Context->OMSetRenderTargets(1, &d3dTexture->RenderTarget, nullptr);
        Context->ClearRenderTargetView(d3dTexture->RenderTarget, info.ClearColor);
    }

    void D3D11CommandList::EndRenderPass()
    {
        // Does nothing.
    }

    void D3D11CommandList::SetViewport(const Viewport& viewport)
    {
        D3D11_VIEWPORT d3dViewport
        {
            .TopLeftX = static_cast<FLOAT>(viewport.X),
            .TopLeftY = static_cast<FLOAT>(viewport.Y),
            .Width = static_cast<FLOAT>(viewport.Width),
            .Height = static_cast<FLOAT>(viewport.Height),
            .MinDepth = viewport.MinDepth,
            .MaxDepth = viewport.MaxDepth
        };

        Context->RSSetViewports(1, &d3dViewport);
    }

    void D3D11CommandList::SetPipeline(Pipeline* pipeline)
    {
        GS_NULL_CHECK(pipeline);
        const auto d3dPipeline = dynamic_cast<D3D11Pipeline*>(pipeline);

        Context->VSSetShader(d3dPipeline->VertexShader, nullptr, 0);
        Context->PSSetShader(d3dPipeline->PixelShader, nullptr, 0);
        Context->IASetInputLayout(d3dPipeline->InputLayout);

        // TODO: Add primitive type to pipeline.
        Context->IASetPrimitiveTopology(D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
    }

    void D3D11CommandList::SetVertexBuffer(uint32_t slot, Buffer* buffer, uint32_t stride, uint32_t offset)
    {
        GS_NULL_CHECK(buffer);
        const auto d3dBuffer = dynamic_cast<D3D11Buffer*>(buffer);

        Context->IASetVertexBuffers(slot, 1, &d3dBuffer->Buffer, &stride, &offset);
    }

    void D3D11CommandList::SetIndexBuffer(Buffer* buffer, Format format)
    {
        GS_NULL_CHECK(buffer);
        const auto d3dBuffer = dynamic_cast<D3D11Buffer*>(buffer);

        DXGI_FORMAT dxgiFormat = Utils::FormatToD3D(format);

        Context->IASetIndexBuffer(d3dBuffer->Buffer, dxgiFormat, 0);
    }

    void D3D11CommandList::DrawIndexed(uint32_t numElements)
    {
        Context->DrawIndexed(numElements, 0, 0);
    }
}

File: src/grabs.Graphics/D3D11/D3D11CommandList.cs
using System.Diagnostics;
using System.Runtime.CompilerServices;
using grabs.Core;
using Vortice.Direct3D;
using Vortice.Direct3D11;
using Vortice.Mathematics;

namespace grabs.Graphics.D3D11;

internal sealed class D3D11CommandList : CommandList
{
    private Dictionary<uint, VertexBufferInfo>? _vertexBuffers;
    
    public const int MaxColorAttachments = 8;
    
    private readonly ID3D11RenderTargetView[] _targetCache;
    
    public readonly ID3D11DeviceContext Context;

    public ID3D11CommandList? CommandList;
    
    public D3D11CommandList(ID3D11Device device)
    {
        _targetCache = new ID3D11RenderTargetView[MaxColorAttachments];
        
        Context = device.CreateDeferredContext();
    }
    
    public override void Begin()
    {
        CommandList?.Dispose();
        CommandList = null;
    }
    
    public override void End()
    {
        Debug.Assert(CommandList == null, "Command List was not null, have you called Begin()?");
        
        CommandList = Context.FinishCommandList(false);
    }
    
    public override void BeginRenderPass(in RenderPassInfo info)
    {
        Debug.Assert(info.ColorAttachments.Length <= MaxColorAttachments);

        int numColorAttachments = info.ColorAttachments.Length;
        for (int i = 0; i < numColorAttachments; i++)
        {
            ref readonly ColorAttachmentInfo attachment = ref info.ColorAttachments[i];
            D3D11Texture d3dTexture = (D3D11Texture) attachment.Texture;
            
            Debug.Assert(d3dTexture.RenderTarget != null, "Render Target is null, has the texture been created as a color attachment?");

            _targetCache[i] = d3dTexture.RenderTarget;
            
            if (attachment.LoadOp == LoadOp.Clear)
            {
                ColorF clearColor = attachment.ClearColor;
                Context.ClearRenderTargetView(d3dTexture.RenderTarget,
                    new Color4(clearColor.R, clearColor.G, clearColor.B, clearColor.A));
            }
        }
        
        Context.OMSetRenderTargets(new ReadOnlySpan<ID3D11RenderTargetView>(_targetCache, 0, numColorAttachments));
    }
    
    public override void EndRenderPass() { }
    
    public override void SetViewport(in Viewport viewport)
    {
        Vortice.Mathematics.Viewport d3dViewport = new Vortice.Mathematics.Viewport()
        {
            X = viewport.X,
            Y = viewport.Y,
            Width = viewport.Width,
            Height = viewport.Height,
            MinDepth = viewport.MinDepth,
            MaxDepth = viewport.MaxDepth
        };
        
        Context.RSSetViewport(d3dViewport);
    }
    
    public override void SetPipeline(Pipeline pipeline)
    {
        D3D11Pipeline d3dPipeline = (D3D11Pipeline) pipeline;
        
        Context.VSSetShader(d3dPipeline.VertexShader);
        Context.PSSetShader(d3dPipeline.PixelShader);
        Context.IASetInputLayout(d3dPipeline.Layout);
        Context.IASetPrimitiveTopology(PrimitiveTopology.TriangleList);

        _vertexBuffers = d3dPipeline.VertexBuffers;
    }
    
    public override void SetVertexBuffer(uint slot, Buffer vertexBuffer, uint offset = 0)
    {
        D3D11Buffer d3dBuffer = (D3D11Buffer) vertexBuffer;
        
        Debug.Assert(_vertexBuffers != null);
        
        Context.IASetVertexBuffer(slot, d3dBuffer.Buffer, _vertexBuffers[slot].Stride, offset);
    }
    
    public override void SetIndexBuffer(Buffer indexBuffer, Format format, uint offset = 0)
    {
        D3D11Buffer d3dBuffer = (D3D11Buffer) indexBuffer;
        
        Context.IASetIndexBuffer(d3dBuffer.Buffer, format.ToD3D(), offset);
    }
    
    public override void Draw(uint numVertices)
    {
        Context.Draw(numVertices, 0);
    }
    
    public override void DrawIndexed(uint numIndices)
    {
        Context.DrawIndexed(numIndices, 0, 0);
    }
    
    public override void Dispose()
    {
        CommandList?.Dispose();
        Context.Dispose();
    }
}